Build pipelines: Add existing pipelines by copying existing ones

suskin opened the pull request:
Related to crossplane/crossplane#1235

We want to have automated build and publishing pipelines for the templating controller repo. Since it's a repository for a go-based controller image, and uses the shared upbound build logic, the easiest starting point is to just copy an existing setup. In this case, we've copied the pipelines from `provider-gcp` and adapted them a little bit.

## Testing

* Here's a test of the tag job: https://jenkinsci.upbound.io/job/crossplane/job/templating-controller/job/templating-controller-tag/job/test-add-existing-pipelines/2/console . Note that it failed because it was not allowed to push to my fork, which is expected.
* Here's a test of the main build job: https://jenkinsci.upbound.io/job/crossplane/job/templating-controller/job/templating-controller/job/add-existing-pipelines/2/console . This mostly shows that the job can go all the way through its steps without failing.
* The first time we use the pipelines after merge will also be a bit of a test, so we'll want to keep an eye on it when we use it for the first time. In testing we exercised most of the pipeline path, but stopped short of publishing an image to the templating-controller docker repo.

suskin committed:
pipelines: add pipelines from provider-gcp

This copies the pipelines from provider-gcp. Since we are bundling a
go-based controller image in that stack repository, and since we are
using the shared `build` repository to do it, we should be able to copy
the build logic from there.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin committed:
Jenkinsfiles: remove build steps which were specific to other repos

Now that we have a copy of the Jenkinsfile from another repo, we want to
remove the parts of it which are not relevant, because our repo doesn't
have them.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin committed:
release: document release process

We're re-using an existing process, so we can just link over to the
existing process's documentation.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

muvaf commented:
This looks good to me so far. Could you also add codecov report upload stage like: https://github.com/crossplane/crossplane/blob/f70bf881/Jenkinsfile#L21 and https://github.com/crossplane/crossplane/blob/f70bf881/Jenkinsfile#L100

@negz  already created a token in codecov. So, instead of `credentials('codecov-crossplane')`, you can use `credentials('codecov-templating-controller')`.

This will let us know about the test coverage on each PR.

suskin commented:
> This looks good to me so far. Could you also add codecov report upload stage like: https://github.com/crossplane/crossplane/blob/f70bf881/Jenkinsfile#L21 and https://github.com/crossplane/crossplane/blob/f70bf881/Jenkinsfile#L100
> 
> @negz already created a token in codecov. So, instead of `credentials('codecov-crossplane')`, you can use `credentials('codecov-templating-controller')`.
> 
> This will let us know about the test coverage on each PR.

This is a bit out of the scope of the goal of this initial PR, since the primary intent is to get *any* pipeline set up, especially for publishing artifacts built from the repository. However, I'll try it out within a timebox, and if I'm able to get it to work pretty easily, I'll include it in this PR.

suskin committed:
build: add cobertura reporting

Now that we have cobertura credentials specific to the
templating-controller project, we can enable cobertura coverage
reporting.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin commented:
Ok, I believe I have the cobertura reporting working. I added some changes, and tested them on Jenkins in a branch. I was going to link to the successful test job execution, but Jenkins already deleted it haha.

I also added our Jenkins's user as a member of the project with `write` permissions, since when I first told Jenkins to run something for this PR it complained about not having write permissions. I think it will probably comment on the PR when it's done, but here's the link in case it doesn't: https://jenkinsci.upbound.io/job/crossplane/job/templating-controller/job/templating-controller/view/change-requests/job/PR-12/1/

displague commented on Jenkinsfile.tag:
```diff
@@ -0,0 +1,42 @@
+        stage('Prepare') {
+            steps {
+                 // github credentials are not setup to push over https in jenkins. add the github token to the url
+                sh "git config remote.origin.url https://${GITHUB_UPBOUND_BOT_USR}:${GITHUB_UPBOUND_BOT_PSW}@\$(git config --get remote.origin.url | sed -e 's/https:\\/\\///')"
```
Perhaps in all of the pipeline that take this approach, we could simplify the authentication requirements by using `git config --global --add url."git@github.com:".insteadOf "https://github.com/"` and then relying on ssh credentials (which I imagine are already available or need less configuration).

suskin replied:
Yeah - I'm thinking it'd be beneficial to share code between the different pipelines first, so that any upgrades we do would apply for all of them

muvaf reviewed (approved):
@suskin Thank you for getting this all figured out, including test report upload!
